apps/dapp-console-api/src/middleware/faucetClaimRateLimiter.ts

```diff
@@ -0,0 +1,45 @@
+import { TRPCError } from '@trpc/server'
+import type { Logger } from 'pino'
+
+import { envVars } from '@/constants'
+import type { GrowthbookStore } from '@/growthbook'
+import { metrics } from '@/monitoring/metrics'
+import type { Trpc } from '@/Trpc'
+import { getFaucetClaimRate } from '@/utils'
+import type { RedisCache } from '@/utils/redis'
+
+/** Middleware used for checking if a user is under the faucet claim rate limit */
+export const faucetClaimRateLimiter = (
+  trpc: Trpc,
+  growthBookStore: GrowthbookStore,
+  redisCache: RedisCache,
+  logger?: Logger,
+) => {
+  return trpc.middleware(async ({ ctx, next }) => {
+    const { ip } = ctx.req
+    console.log(
+      'enable_faucet_rate_limit:',
+      growthBookStore.get('enable_faucet_rate_limit'),
+    )
+    if (!growthBookStore.get('enable_faucet_rate_limit')) {
+      return next()
+    }
+
+    if (!ip) {
+      logger?.warn('faucet encountered a request with an undefined ip address')
+      return next()
+    }
+
+    const requestCount = await getFaucetClaimRate(redisCache, ip)
```

Comment:
Shouldn't we be incrementing the claim rate here? `INCR` updates and returns the latest value for the counter when updated. With this pattern we no longer need the `update`.

Simply something like
```javascript
const count = await incIPClaimRate(redisCache, ip)
if count > BigInt(envVars.FAUCET_RATE_LIMIT) {
   // rate limit
}
```

Reply:
I intentionally am not incrementing here. we want to increment after a successful claim, so that is why I am doing the increment inside the route. I do not think they should be incremented strictly for hitting this endpoint, because there is a chance that the claim errors out due to an issue on our side and the user needs to retry it.

Reply:
Event if it's a retry, the same RPC should be rate limited right? If there's transient error even on our side, continually spamming the retry will just make the problem worse. The client should back off a retry according to the overall rate limit on the endpoint

Reply:
This should protect us from transient errors on our side. A claim is considered "successful" as soon as the tx hash is generated but before the transaction is confirmed on chain. This is where we are seeing issues, user's are continuing to retry to claim from the faucet while their claim tx is pending and has not gone through yet. By incrementing the rate limit here it allows us to know that the claim tx was created successfully and therefore we can more heavily rate limit this endpoint. Also something to keep in mind is we have a top level rate limiter that protects requests across all of our backend's endpoints.

Reply:
I see so the rate limit is on the txs created not the requests.

Curious why not just rate limit the top level requests for making the claim?
> This is where we are seeing issues, user's are continuing to retry to claim from the faucet while their claim tx is pending and has not gone through yet

This right here feels like the problem. We're only rate limiting when the claim goes through but will let callers attempt to create as many claim txs as it takes which will make the problem worse when overloaded. The top level rpc should instead be rate limited to begin this. 

In both scenarios, it's the top level call that should fail
1. Prolonged confirmation on-chain. Users increase rpc count for a successful claims -> rate limit
2. User tries to claim too much -> rate limite

Reply:
> We're only rate limiting when the claim goes through but will let callers attempt to create as many claim txs as it takes which will make the problem worse when overloaded.

With the way this is currently implemented the rate limit will be incremented as soon as the tx is created. So in the case of prolonged confirmation on-chain we are still covered. 

I prefer to do the incrementation here instead of at the top level so that we can have a more strict rate limit, since we know that the tx has been created and is just waiting for confirmation on chain rather than a top level rate limit where an error might be due to something else and a retry could be valid.

Reply:
okay sounds good